util/coverage/coverage_off_target.py

```diff
@@ -0,0 +1,253 @@
+
+
+def generate_report(out_dir: Path, merged_profile: Path, merged_so: Path,
+                    print_text_report: bool):
+    """Generate a coverage report.
+
+    This function generates a coverage report from the given merged profile and shared
+    library.
+
+    Args:
+        out_dir: Path where to save the html report.
+        merged_profile: Path of merged profile data.
+        merged_so: Path of the shared library with coverage data.
+        print_text_report: Whether to print the text report.
+    """
+    # Generate html coverage report.
+    run(LLVM_COV, "show", "-show-line-counts", "-show-regions",
+        f"--project-title={REPORT_TITLE}", "--format=html",
+        f"--output-dir=./{out_dir}", "--instr-profile", str(merged_profile),
+        str(merged_so))
+    # Generate text coverage report.
+    with (out_dir / "report.txt").open("w") as f:
+        f.write("\n".join(
+            run(LLVM_COV, "report", "--instr-profile", str(merged_profile),
+                str(merged_so))))
+    if print_text_report:
+        print("\n".join(
+            run(LLVM_COV, "report", "--use-color", "--instr-profile",
+                str(merged_profile), str(merged_so))))
+    print(f"Saved coverage artifacts in {out_dir}.")
+
+
+def create_out_dir(out_root_dir: Path):
+    """Create the directory for coverage report.
+
+    The report will be saved in `out_root_dir`/<HEAD_TIMESTAMP>-<HEAD_HASH>/unittest/`.
+
+    Args:
+        out_root_dir: Root of the output directory.
+
+    Returns:
+        Path where to save the coverage report.
+    """
+    # Report artifacts are placed under `out_root_dir`/<head_timestamp>-<head_hash>/unittest/.
+    [head_hash] = run("git", "rev-parse", "HEAD")
+    [head_timestamp] = run("git", "show", "-s", "--format=%ct", "HEAD")
+    out_dir = out_root_dir / f"{head_timestamp}-{head_hash}/unittest/"
+    out_dir.mkdir(parents=True, exist_ok=True)
+    return out_dir
+
+
+@app.command()
+def main(log_level: LogLevel = LogLevel.none,
+         out_root_dir: Path = Path("./coverage"),
+         print_text_report: bool = False):
+    """Measures unit test coverage of silicon creator code."""
+    if log_level != LogLevel.none:
+        log.basicConfig(level=log.getLevelName(log_level.upper()))
+
+    out_dir = create_out_dir(out_root_dir)
+    # Merged profile data path
+    merged_profile = out_dir / "merged.profdata"
+    # Merged shared library path with all coverage data we want
+    merged_so = out_dir / "merged.so"
+
+    # Build device libs and create a merged shared library to avoid hash-mismatch
+    # warnings related to unused inline functions.
+    device_libs = gather_and_build_libs()
+    create_merged_library(device_libs, merged_so)
+    # Coverage report is generated from merged profile data from unit tests and merged
+    # coverage data from the combined shared library.
+    measure_unit_test_coverage(merged_profile)
+    generate_report(out_dir, merged_profile, merged_so, print_text_report)
```

Comment:
Have you thought about making this script into a Bazel rule? You could probably factor out the `bazel query` operations into `genquery()` targets in a BUILD file.

Reply:
Not really but added your comment to #16761

Reply:
Resolving now since this can be handled later.